### PR TITLE
[CI] Jenkins package pipeline for agentbeat

### PR DESCRIPTION
**Pull request by pazone**
## Proposed commit message
Adds agentbeat to Jenkins packaging pipeline to fix the [legacy DRA publishing](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fpackaging/detail/main/4000/pipeline/)

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

- Relates https://github.com/elastic/ingest-dev/issues/3151

### CODE REVIEW

File: .ci/packaging.groovy [hunk 1]
```diff
@@ -280,6 +281,9 @@ def generateLinuxStep(beat) {
       withEnv(["HOME=${env.WORKSPACE}", "PLATFORMS=${linuxPlatforms()}", "BEATS_FOLDER=${beat}"]) {
         withGithubNotify(context: "Packaging Linux ${beat}") {
           deleteDir()
+          if (beat.equals('x-pack/agentbeat')) {
```
Comment (blakerouse) [inline .ci/packaging.groovy:284]: This needs to be installed for osquerybeat as well.

**Review by blakerouse — APPROVED** [review]
Looks good!

File: .buildkite/scripts/generate_xpack_osquerybeat_pipeline.sh [hunk 2]
```diff
@@ -181,7 +181,10 @@ if are_conditions_met_packaging; then
     steps:
       - label: ":linux: Packaging Linux"
         key: "packaging-linux"
-        command: "cd $BEATS_PROJECT_NAME && mage package"
+        command: |
+          sudo apt-get update -y
+          DEBIAN_FRONTEND=noninteractive sudo apt-get install --no-install-recommends --yes msitools
```
Comment (v1v) [inline .buildkite/scripts/generate_xpack_osquerybeat_pipeline.sh:186]: what's the reason for not using the existing script?

```suggestion
          .ci/scripts/install-msitools.sh
```

File: x-pack/osquerybeat/internal/distro/distro.go [hunk 3]
```diff
@@ -178,7 +178,7 @@ var specs = map[OSArch]Spec{
 	{"linux", "arm64"}:   {"_1.linux_aarch64.tar.gz", osqueryDistroLinuxARMSHA256, true},
 	{"darwin", "amd64"}:  {osqueryPkgExt, osqueryDistroDarwinSHA256, true},
 	{"darwin", "arm64"}:  {osqueryPkgExt, osqueryDistroDarwinSHA256, true},
-	{"windows", "amd64"}: {osqueryMSIExt, osqueryDistroWindowsSHA256, false},
+	{"windows", "amd64"}: {osqueryMSIExt, osqueryDistroWindowsSHA256, true},
```
Comment (v1v) [inline x-pack/osquerybeat/internal/distro/distro.go:181]: what's the reason for this change? the title says something about agentbeat only, rather than osquery?

File: x-pack/agentbeat/Jenkinsfile.yml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+            changeset:         ## when PR contains any of those entries in the changeset
+              - "^x-pack/agentbeat/.*"
+              - "^auditbeat/.*"
+              - "^filebeat/.*"
+              - "^heartbeat/.*"
+              - "^metricbeat/.*"
+              - "^osquerybeat/.*"
+              - "^packetbeat/.*"
+              - "@oss"               ## special token regarding the changeset for the oss
+              - "@xpack"
```
Comment (v1v) [inline x-pack/agentbeat/Jenkinsfile.yml:33]: The changeset at the stage level should be a subset of the changeset in the top-level `when`. 

Can you add these changeset values in `when.changeset`?

File: .buildkite/scripts/generate_xpack_osquerybeat_pipeline.sh (hunk 2 above)
Comment (pazone) [reply .buildkite/scripts/generate_xpack_osquerybeat_pipeline.sh:186]: We are trying not to use the `.ci` scripts in buildkite due to migration. Because it is going to be removed by EoM.

File: x-pack/osquerybeat/internal/distro/distro.go (hunk 3 above)
Comment (pazone) [reply x-pack/osquerybeat/internal/distro/distro.go:181]: Agentbeat contains requires osquerybeat and other beats. It's required because we unpack the msi package

File: .buildkite/scripts/generate_xpack_osquerybeat_pipeline.sh (hunk 2 above)
Comment (v1v) [reply .buildkite/scripts/generate_xpack_osquerybeat_pipeline.sh:186]: what's the reason you don't use `.buildkite/scripts` and then apply the changes in the `Jenkinsfile`?

Comment (pazone) [reply .buildkite/scripts/generate_xpack_osquerybeat_pipeline.sh:186]: If I understand the question correctly, the `.buildkite/scripts` changes are related to the osquerybeat packaging in bk

Comment (v1v) [reply .buildkite/scripts/generate_xpack_osquerybeat_pipeline.sh:186]: sorry if I was vague, I meant what's the reason you don't rename `.ci/scripts/install-msitools.sh` for `.buildkite scripts/install-msitools.sh` and then use it in both places (Jenkins and Buildkite pipelines).

IIUC, `.ci` will be removed but `.buildkite` will remain, therefore you can take advantage of removing the duplicated code for `apt-get update` and `apt-get install msitools` using the `.buildkite/scripts/install-msitools.sh` and use that script in both CI pipelines.

File: x-pack/agentbeat/Jenkinsfile.yml (hunk 4 above)
Comment (pazone) [reply x-pack/agentbeat/Jenkinsfile.yml:39]: Done

File: .buildkite/scripts/generate_xpack_osquerybeat_pipeline.sh (hunk 2 above)
Comment (pazone) [reply .buildkite/scripts/generate_xpack_osquerybeat_pipeline.sh]: Ok. I updated the branch and now this script is deleted

**Review by rdner — APPROVED** [review]

File: .ci/scripts/install-msitools.sh [hunk 5]
```diff
@@ -0,0 +1,4 @@
+set -euo pipefail
```
Comment (v1v) [inline .ci/scripts/install-msitools.sh:1]: IIUC, we said `.ci/scripts/install-msitools.sh` will be renamed to `.buildkite/scripts/install-msitools.sh` and then all the references for `.ci/scripts/install-msitools.sh` will be changed for `.buildkite/scripts/install-msitools.sh`, similarly done for `.buildkite/x-pack/pipeline.xpack.osquerybeat.yml` that will use `.buildkite/scripts/install-msitools.sh`

File: .ci/packaging.groovy [hunk 6]
```diff
@@ -280,6 +281,9 @@ def generateLinuxStep(beat) {
       withEnv(["HOME=${env.WORKSPACE}", "PLATFORMS=${linuxPlatforms()}", "BEATS_FOLDER=${beat}"]) {
         withGithubNotify(context: "Packaging Linux ${beat}") {
           deleteDir()
+          if (beat.equals('x-pack/agentbeat') || beat.equals('x-pack/osquerybeat')) {
+            sh(label: 'install msitools', script: '.ci/scripts/install-msitools.sh')
```
Comment (v1v) [inline .ci/packaging.groovy:285]:
```suggestion
            sh(label: 'install msitools', script: '.buildkite/scripts/install-msitools.sh')
```

File: Jenkinsfile [hunk 7]
```diff
@@ -608,6 +608,9 @@ def targetWithoutNode(Map args = [:]) {
         }
       }
       withTools(k8s: installK8s, gcp: withGCP, nodejs: withNodejs) {
+        if (isPackaging && (directory.equals('x-pack/agentbeat') || directory.equals('x-pack/osquerybeat'))) {
+          sh(label: 'install msitools', script: '.ci/scripts/install-msitools.sh')
```
Comment (v1v) [inline Jenkinsfile:612]:
```suggestion
          sh(label: 'install msitools', script: '.buildkite/scripts/install-msitools.sh')
```

File: x-pack/agentbeat/Jenkinsfile.yml [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
```
[the remark this answers is not in the document]
Comment (pazone) [reply x-pack/agentbeat/Jenkinsfile.yml:39]: Decided that we leave it as it is for now

File: .buildkite/scripts/generate_xpack_osquerybeat_pipeline.sh (hunk 2 above)
Comment (pazone) [reply .buildkite/scripts/generate_xpack_osquerybeat_pipeline.sh]: We will add the `msitools` to the base image and delete all install scripts after that

Comment (pazone) [reply .buildkite/scripts/generate_xpack_osquerybeat_pipeline.sh]: Raised a [follow up issue](https://github.com/elastic/ingest-dev/issues/3194) for that

**Review by dliappis — APPROVED** [review]
LGTM provided that CI is green and the corresponding packaging step on Jenkins works as expected.

Regarding the use of a script (under `.buildkite/`) to install `msitools`, I also had the same initial thought, but on a second thought, we could simply pre-install it on our base image; this is tracked in https://github.com/elastic/ingest-dev/issues/3194

**Comment by pazone** [comment]
@v1v Moved the install script to `.buildkite/scripts`

**Comment by pazone** [comment]
Now packaging [works](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats/detail/PR-38951/22/pipeline/). The PR is ready to merge

**Comment by cmacknz** [comment]
Force merging to fix the 8.14 SNAPSHOT build

**Comment by cmacknz** [comment]
I think something is still broken in the DRA packaging logic:

```
[2024-04-17T01:56:48.997Z] Started by user Craig MacKenzie

[2024-04-17T01:56:49.074Z] Connecting to https://api.github.com/ using Jenkins - beats-ci

[2024-04-17T01:56:49.571Z] Obtained .ci/packaging.groovy from 3f06575996bc91402a5c330c18b0e7451c43785b

[2024-04-17T01:56:49.571Z] Resume disabled by user, switching to high-performance, low-durability mode.

[2024-04-17T01:56:49.587Z] org.codehaus.groovy.control.MultipleCompilationErrorsException: startup failed:

[2024-04-17T01:56:49.587Z] WorkflowScript: 208: expecting ']', found 'x-pack/auditbeat' @ line 208, column 5.

[2024-04-17T01:56:49.587Z]        'x-pack/auditbeat',

[2024-04-17T01:56:49.587Z]        ^

[2024-04-17T01:56:49.587Z] 

[2024-04-17T01:56:49.587Z] 1 error

[2024-04-17T01:56:49.587Z] 

[2024-04-17T01:56:49.587Z] 	at org.codehaus.groovy.control.ErrorCollector.failIfErrors(ErrorCollector.java:309)

[2024-04-17T01:56:49.587Z] 	at org.codehaus.groovy.control.ErrorCollector.addFatalError(ErrorCollector.java:149)

[2024-04-17T01:56:49.587Z] 	at org.codehaus.groovy.control.ErrorCollector.addError(ErrorCollector.java:119)

[2024-04-17T01:56:49.587Z] 	at org.codehaus.groovy.control.ErrorCollector.addError(ErrorCollector.java:131)

[2024-04-17T01:56:49.587Z] 	at org.codehaus.groovy.control.SourceUnit.addError(SourceUnit.java:349)

[2024-04-17T01:56:49.587Z] 	at org.codehaus.groovy.antlr.AntlrParserPlugin.transformCSTIntoAST(AntlrParserPlugin.java:225)

[2024-04-17T01:56:49.587Z] 	at org.codehaus.groovy.antlr.AntlrParserPlugin.parseCST(AntlrParserPlugin.java:191)

[2024-04-17T01:56:49.587Z] 	at org.codehaus.groovy.control.SourceUnit.parse(SourceUnit.java:233)

[2024-04-17T01:56:49.587Z] 	at org.codehaus.groovy.control.CompilationUnit$1.call(CompilationUnit.java:189)

[2024-04-17T01:56:49.587Z] 	at org.codehaus.groovy.control.CompilationUnit.applyToSourceUnits(CompilationUnit.java:966)

[2024-04-17T01:56:49.587Z] 	at org.codehaus.groovy.control.CompilationUnit.doPhaseOperation(CompilationUnit.java:626)

[2024-04-17T01:56:49.587Z] 	at org.codehaus.groovy.control.CompilationUnit.processPhaseOperations(CompilationUnit.java:602)

[2024-04-17T01:56:49.587Z] 	at org.codehaus.groovy.control.CompilationUnit.compile(CompilationUnit.java:579)

[2024-04-17T01:56:49.587Z] 	at groovy.lang.GroovyClassLoader.doParseClass(GroovyClassLoader.java:323)

[2024-04-17T01:56:49.587Z] 	at groovy.lang.GroovyClassLoader.parseClass(GroovyClassLoader.java:293)

[2024-04-17T01:56:49.587Z] 	at groovy.lang.GroovyShell.parseClass(GroovyShell.java:677)

[2024-04-17T01:56:49.587Z] 	at groovy.lang.GroovyShell.parse(GroovyShell.java:689)

[2024-04-17T01:56:49.587Z] 	at org.jenkinsci.plugins.workflow.cps.CpsGroovyShell.doParse(CpsGroovyShell.java:142)

[2024-04-17T01:56:49.587Z] 	at org.jenkinsci.plugins.workflow.cps.CpsGroovyShell.reparse(CpsGroovyShell.java:127)

[2024-04-17T01:56:49.587Z] 	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.parseScript(CpsFlowExecution.java:561)

[2024-04-17T01:56:49.587Z] 	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.start(CpsFlowExecution.java:513)

[2024-04-17T01:56:49.587Z] 	at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:335)

[2024-04-17T01:56:49.587Z] 	at hudson.model.ResourceController.execute(ResourceController.java:101)

[2024-04-17T01:56:49.587Z] 	at hudson.model.Executor.run(Executor.java:442)

[2024-04-17T01:56:49.599Z] Finished: FAILURE
```

**Comment by dliappis** [comment]
@pazone see https://github.com/elastic/beats/pull/38951#issuecomment-2060200712. This PR will also need to be backported I guess.

**Comment by pierrehilbert** [comment]
FYI https://github.com/elastic/beats/issues/39000
